xtractmime/main.py

```diff
@@ -0,0 +1,78 @@
+from typing import List, Union, Tuple
+
+_APACHE_TYPES = [
+    b"text/plain",
+    b"text/plain; charset=ISO-8859-1",
+    b"text/plain; charset=iso-8859-1",
+    b"text/plain; charset=UTF-8",
+]
+
+
+_WHITESPACE_BYTES = {b"\t", b"\r", b"\x0c", b"\n", b" "}
+
+
+# handling resource metadata
+def _should_check_for_apache_bug(supplied_type, http_origin):
+    return http_origin and supplied_type in _APACHE_TYPES
+
+
+# reading resource header
+def _read_resource_header(body):
+    buffer = memoryview(body)
+    if len(body) < 1445:
+        return buffer
+    else:
+        return buffer[:1445]
+    
+
+
+# Matching a MIME type pattern
+def _is_match_mime_pattern(
+    input_bytes, byte_pattern, pattern_mask, lead_whitespace=None
+):
+    input_size, pattern_size, mask_size = (
+        len(input_bytes),
+        len(byte_pattern),
+        len(pattern_mask),
+    )
+
+    if pattern_size != mask_size:
+        raise ValueError("pattern's length should match mask's length")
+
+    if input_size < pattern_size:
+        return False
+
+    input_index, pattern_index = 0, 0
+
+    if lead_whitespace:
+        while (
+            input_index < input_size
+            and input_bytes[input_index:input_index+1] in _WHITESPACE_BYTES
+        ):
+            input_index += 1
+
+    while pattern_index < pattern_size:
+        masked_byte = bytes([input_bytes[input_index] & pattern_mask[pattern_index]])
+        if masked_byte != byte_pattern[pattern_index:pattern_index+1]:
+            return False
+        input_index += 1
+        pattern_index += 1
+
+    return True
+
+
+# main function
+def extract_mime(
+    body: bytes,
+    *,
+    content_types: List[bytes] = [],
```

Comment:
Please see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments.

This also applies to `extra_types`

Reply:
@Gallaecio mentioned about this earlier [here](https://github.com/scrapy/xtractmime/pull/2#discussion_r648657242) that there is no need for it as we are not editing `content-types` and `extra_types` within the function.

Reply:
Right, in any case, I think it's better not to leave either occurrence as a list. It's not hard to handle:

`content_types: Optional[List[bytes]] = None,`

and later

`content_types = content_types or []` or `content_types = content_types if content_types is not None else []`

Reply:
Thanks, I will change them accordingly.